Return of the Banman

dongcarl opened the pull request:
**Old English à la Beowulf**
```
Banman wæs bréme    --blaéd wíde sprang--
Connmanes eafera    Coreum in.
aéglaéca            léodum forstandan
Swá bealdode        bearn Connmanes
guma gúðum cúð      gódum daédum·
dréah æfter dóme·   nealles druncne slóg
```

**Modern English Translation**
```
Banman was famed              --his renown spread wide--
Conman's hier,                in Core-land.
against the evil creature     defend the people
Thus he was bold,             the son of Connman
man famed in war,             for good deeds;
he led his life for glory,    never, having drunk, slew
```

--

With @theuni's blessing, here is Banman, rebased. Original PR: https://github.com/bitcoin/bitcoin/pull/11457

--

Followup PRs:
1. Give `CNode` a `Disconnect` method ([source](https://github.com/bitcoin/bitcoin/pull/14605#discussion_r248065847))
2. Add a comment to `std::atomic_bool fDisconnect` in `net.h` that setting this to true will cause the node to be disconnected the next time `DisconnectNodes()` runs ([source](https://github.com/bitcoin/bitcoin/pull/14605#discussion_r248384309))

theuni committed:
net: Break disconnecting out of Ban()

These are separate events which need to be carried out by separate subsystems.

This also cleans up some whitespace and tabs in qt to avoid getting flagged by
the linter.

Current behavior is preserved.

theuni committed:
tests: remove member connman/peerLogic in TestingSetup

theuni committed:
net: split up addresses/ban dumps in preparation for moving them

theuni committed:
banman: create and split out banman

Some say he has always been.

theuni committed:
banman: pass the banfile path in

There's no need to hard-code the path here. Passing it in means that there are
no ordering concerns wrt establishing the datadir.

theuni committed:
banman: pass in default ban time as a parameter

Removes the dependency on arg parsing.

theuni committed:
net: move BanMan to its own files

theuni committed:
scripted-diff: batch-rename BanMan members

-BEGIN VERIFY SCRIPT-
sed -i "s/clientInterface/m_client_interface/g" src/banman.h src/banman.cpp
sed -i "s/setBannedIsDirty/m_is_dirty/g" src/banman.h src/banman.cpp
sed -i "s/cs_setBanned/m_cs_banned/g" src/banman.h src/banman.cpp
sed -i "s/setBanned/m_banned/g" src/banman.h src/banman.cpp
-END VERIFY SCRIPT-

theuni committed:
banman: add thread annotations and mark members const where possible

Also remove misleading comment. ClearBanned is used by rpc as well.

MarcoFalke commented on src/qt/rpcconsole.cpp:
```diff
@@ -1221,16 +1221,17 @@ void RPCConsole::banSelectedNode(int bantime)
-	if(stats) {
+    // Find possible nodes, ban it and clear the selected node
+    const CNodeCombinedStats *stats = clientModel->getPeerTableModel()->getNodeStats(detailNodeRow);
+    if(stats) {
```
Accidental changes in the first commit?

practicalswift commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+#include <netaddress.h>
+#include <utiltime.h>
+#include <util.h>
+#include <ui_interface.h>
```
Nit: Sort includes :-)

practicalswift commented on src/addrdb.cpp:
```diff
@@ -105,9 +105,8 @@ bool DeserializeFileDB(const fs::path& path, Data& data)
 }
 
-CBanDB::CBanDB()
+CBanDB::CBanDB(fs::path ban_file) : pathBanlist(std::move(ban_file))
```
Make sure parameter name matches between declaration and definition :-)

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+public:
+    ~BanMan();
+    BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
+    void Ban(const CNetAddr& netAddr, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
```
Make sure parameter names match between:

```
    void Ban(const CNetAddr& netAddr, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
    void BanMan::Ban(const CNetAddr& addr, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch) {
```

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    ~BanMan();
+    BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time);
+    void Ban(const CNetAddr& netAddr, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
+    void Ban(const CSubNet& subNet, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
```
Make sure these match:

```
    void Ban(const CSubNet& subNet, const BanReason& reason, int64_t bantimeoffset = 0, bool sinceUnixEpoch = false);
    void BanMan::Ban(const CSubNet& subNet, const BanReason &banReason, int64_t bantimeoffset, bool sinceUnixEpoch) {
```

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    void ClearBanned();
+    bool IsBanned(CNetAddr ip);
+    bool IsBanned(CSubNet subnet);
+    bool Unban(const CNetAddr &ip);
```
Make sure these match:

```
    bool Unban(const CNetAddr &ip);
    bool BanMan::Unban(const CNetAddr &addr) {
```

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    bool IsBanned(CNetAddr ip);
+    bool IsBanned(CSubNet subnet);
+    bool Unban(const CNetAddr &ip);
+    bool Unban(const CSubNet &ip);
```
Same here:

```
    bool Unban(const CSubNet &ip);
    bool BanMan::Unban(const CSubNet &subNet) {
```

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    bool IsBanned(CSubNet subnet);
+    bool Unban(const CNetAddr &ip);
+    bool Unban(const CSubNet &ip);
+    void GetBanned(banmap_t &banmap);
```
Same here:

```
    void GetBanned(banmap_t &banmap);
    void BanMan::GetBanned(banmap_t &banMap)
```

practicalswift commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    void DumpBanlist();
+
+private:
+    void SetBanned(const banmap_t &banmap);
```
And here :-)

```
    void SetBanned(const banmap_t &banmap);
    void BanMan::SetBanned(const banmap_t &banMap)
```

practicalswift commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+bool BanMan::Unban(const CSubNet &subNet) {
+    {
+        LOCK(m_cs_banned);
+        if (!m_banned.erase(subNet))
```
Nit: `if (m_banned.erase(subNet) == 0)` to increase readability and avoid implicit conversion?

practicalswift commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+        m_is_dirty = true;
+    }
+    DumpBanlist(); //store banlist to disk
+    if(m_client_interface)
```
Since this is new code - please run all new C++ files through `clang-format` (using the project local settings) to get the expected formatting.

practicalswift commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+    if (bantimeoffset <= 0)
+    {
+        bantimeoffset = m_default_ban_time;
+        sinceUnixEpoch = false;
```
Could this be re-formulated to avoid reassigning parameters?

dongcarl replied:
See: https://github.com/bitcoin/bitcoin/pull/14605/commits/5dac8bf9616749b832ce14ccace4a2ab6eaf187b

dongcarl replied:
Reformulated in https://github.com/bitcoin/bitcoin/pull/14605/commits/36e6b241d0ed6f6d6ab503d61dfd95053202ce9b, let me know if you think that's good

practicalswift replied:
@dongcarl Nice! Thanks! :-)

dongcarl committed:
banman: reformulate nBanUtil calculation

Avoid reassigning parameters.

dongcarl committed:
scripted-diff: batch-recase BanMan variables

-BEGIN VERIFY SCRIPT-
sed -i "s/banMap/banmap/g" src/banman.h src/banman.cpp
sed -i "s/netAddr/net_addr/g" src/banman.h src/banman.cpp
sed -i "s/sinceUnixEpoch/since_unix_epoch/g" src/banman.h src/banman.cpp
sed -i "s/bantimeoffset/ban_time_offset/g" src/banman.h src/banman.cpp
sed -i "s/subNet/sub_net/g" src/banman.h src/banman.cpp
sed -i "s/banReason/ban_reason/g" src/banman.h src/banman.cpp
sed -i "s/notifyUI/notify_ui/g" src/banman.h src/banman.cpp
sed -i "s/banEntry/ban_entry/g" src/banman.h src/banman.cpp
sed -i "s/nStart/n_start/g" src/banman.h src/banman.cpp
-END VERIFY SCRIPT-

laanwj commented:
thank you for picking this up!

laanwj commented:
No merge conflicts, but it does need rebase; current build fails wiith
```c++
/.../bitcoin/src/banman.cpp:10:10: fatal error: util.h: No such file or directory
 #include <util.h>
          ^~~~~~~~
compilation terminated.
```

dongcarl commented:
Done.

dongcarl commented:
Rebased, and addressed casing concerns. Additional review welcome.

ryanofsky commented:
I'd like to review this soon, but is there a description of what high level goal of this PR is? Is it cleanup geared towards making code maintenance easier? Or is it intended to enable a new feature? Is it cleaning up technical debt from a previously added feature? The previous PR #14605 just says it "breaks the ban/unban functions out of CConnman and into a new class because, while logically bans are tied to connections, they're really just entries in a database," which I think justifies the change but doesn't really motivate it.

theuni commented on src/banman.cpp:
```diff
@@ -106,11 +106,14 @@ void BanMan::Ban(const CSubNet& subNet, const BanReason& banReason, int64_t bant
     banEntry.banReason = banReason;
+
+    int64_t normalized_bantimeoffset = bantimeoffset;
+    int64_t normalized_sinceUnixEpoch = sinceUnixEpoch;
```
normalized_sinceUnixEpoch should be a bool.

theuni commented on src/banman.cpp:
```diff
@@ -25,7 +25,7 @@ BanMan::BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t
         SweepBanned();            // sweep out unused entries
 
         LogPrint(BCLog::NET, "Loaded %d banned node ips/subnets from banlist.dat  %dms\n",
-            banmap.size(), GetTimeMillis() - nStart);
```
Hungarian or snake_case, pick one :)

dongcarl replied:
Fixed in 9bad9fa1168ec53e92bb8b2df0b5e9d578344a2f

theuni commented on src/banman.cpp:
```diff
@@ -69,78 +69,78 @@ void BanMan::ClearBanned()
-        normalized_bantimeoffset = m_default_ban_time;
-        normalized_sinceUnixEpoch = false;
+    CBanEntry ban_entry(GetTime());
+    ban_entry.banReason = ban_reason;
```
Please don't use special-cases in scripted-diffs to negate part of a change, as that means that manual review is required.

Instead, please add a commit before the scripted-diff which renames any necessary conflicts.

dongcarl replied:
Do you mean rename `CBanEntry.banReason` to `CBanEntry.m_ban_reason` before the scripted diff?

theuni reviewed (commented):
@dongcarl Thanks for picking this up! Added a few nits.

theuni commented:
@ryanofsky This is one step (one of the big ones) closer to instantiated CConnman (self-contained and no-globals), which would allow for all kinds of cool things, primarily running separate net instances against each-other.

Another nice side-effect of the above is that once self-contained, bitcoind could be run (or even compiled!) without net support.

dongcarl committed:
banman: Add, use CBanEntry ctor that takes ban reason

dongcarl commented:
Rebased and addressed reviews!

Empact commented on src/addrdb.h:
```diff
@@ -43,6 +43,11 @@ class CBanEntry
 
+    explicit CBanEntry(int64_t n_create_time_in, BanReason ban_reason_in) : CBanEntry(n_create_time_in)
+    {
+        banReason = ban_reason_in;
```
nit: could be initializer

dongcarl replied:
I don't think this can be initializer as we need to `SetNull()`, the current one does the delegation thing.

Empact commented on src/addrdb.h:
```diff
@@ -92,9 +97,9 @@ class CAddrDB
 {
 private:
-    fs::path pathBanlist;
+    fs::path m_ban_list_path;
```
nit: could be const

Empact commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+#include <util/time.h>
+
+
+BanMan::BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time) : m_client_interface(client_interface), m_ban_db(std::move(ban_file)), m_default_ban_time(default_ban_time)
```
nit: could wrap this line

Empact commented on src/banman.cpp:
```diff
@@ -0,0 +1,202 @@
+BanMan::BanMan(fs::path ban_file, CClientUIInterface* client_interface, int64_t default_ban_time) : m_client_interface(client_interface), m_ban_db(std::move(ban_file)), m_default_ban_time(default_ban_time)
+{
+    if (m_client_interface)
+        m_client_interface->InitMessage(_("Loading banlist..."));
```
nit: one line or braces

Empact commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    bool m_is_dirty GUARDED_BY(m_cs_banned);
+    CClientUIInterface* m_client_interface = nullptr;
+    CBanDB m_ban_db;
+    const int64_t m_default_ban_time;
```
nit: `#include <cstdint>`

Empact commented on src/banman.h:
```diff
@@ -0,0 +1,66 @@
+    const int64_t m_default_ban_time;
+};
+
+extern std::unique_ptr<BanMan> g_banman;
```
nit: `#include <memory>`

Empact commented on src/net.cpp:
```diff
@@ -2668,6 +2473,25 @@ bool CConnman::DisconnectNode(const std::string& strNode)
+    bool disconnected = false;
+    LOCK(cs_vNodes);
+    for (CNode* pnode : vNodes) {
+        if (subnet.Match(static_cast<CNetAddr>(pnode->addr))) {
```
Seems `dynamic_cast` would be more appropriate here, given the cast to base class.

laanwj replied:
I'd advise against using `dynamic_cast`. It's not used anywhere else in the code and it relies on C++ run-time type information which has at least historically a large overhead. I'm also fairly sure it's unnecessary here.

Empact commented on src/net_processing.cpp:
```diff
@@ -2961,14 +2961,16 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman, bool en
+        else if (pnode->addr.IsLocal()) {
+            // Disconnect but don't ban _this_ local node
+            LogPrintf("Warning: disconnecting but not banning local peer %s!\n", pnode->addr.ToString());
+            connman->DisconnectNode(pnode->GetId());
```
nit: could have a `CNode*` version of `DisconnectNode` to avoid the lookup

dongcarl replied:
I'm going to revert to doing
```
pnode->fDisconnect = true;
```
so as to avoid the lookup, then in a followup PR we can make `fDisconnect` private, and have a `Disconnect` method on `CNode`.

laanwj commented:
We want to have this in for 0.18.0, there's not *that* much time anymore.
This has had a fair amount of review.
Let's rebase, address the nits, do a last review round, and merge it.

dongcarl commented:
@laanwj Yes. I will make it happen this week.

dongcarl commented:
Rebased and addressed nits.

laanwj commented:
Thank you @dongkarl!

laanwj commented:
utACK cb1ee04b6a7d56834a6447711ca5500fc8bd83cd

promag commented on src/net_processing.cpp:
```diff
@@ -2962,13 +2962,14 @@ static bool SendRejectsAndCheckIfBanned(CNode* pnode, CConnman* connman, bool en
+            } else {
+                // Disconnect and ban all nodes sharing the address
                 connman->Ban(pnode->addr, BanReasonNodeMisbehaving);
             }
```
Commit b4ca7e2

Discussed with @dongcarl about not changing behavior in this commit, the conclusion is to call `conman->DisconnectNode(pnode->addr)` here and add `pnode->fDisconnect = true;` above.

Sjors commented on src/net.cpp:
```diff
@@ -41,8 +41,11 @@
-// Dump addresses to peers.dat and banlist.dat every 15 minutes (900s)
-#define DUMP_ADDRESSES_INTERVAL 900
+// Dump addresses to peers.dat every 15 minutes (900s)
+static constexpr int DUMP_PEERS_INTERVAL = 60 * 15;
```
Nit: `15 * 60` to be consistent with other `_INTERVAL` constants.

dongcarl replied:
Done!

Sjors commented on src/interfaces/node.cpp:
```diff
@@ -120,24 +120,24 @@ class NodeImpl : public Node
     {
-        if (g_connman) {
-            g_connman->GetBanned(banmap);
+        if (g_banman) {
```
cc @ryanofsky

Sjors reviewed (approved):
tACK 95b495b9697502b2b478dda8cece0ae9ffb6818f

@ryanofsky wrote:
> is there a description of what high level goal of this PR is? Is it cleanup geared towards making code maintenance easier? [...]

I think reducing the 3000 lines in net.cpp by 200 is almost worth it by itself :-)

From the original description in #11457:
> This also makes testing easier as different implementations can be dropped in.

So this is a step towards being able to mock network code and e.g. test banning behavior for a large number of simulated nodes.

dongcarl commented on src/net.cpp:
```diff
@@ -2643,6 +2636,25 @@ bool CConnman::DisconnectNode(const std::string& strNode)
+    LOCK(cs_vNodes);
+    for (CNode* pnode : vNodes) {
+        if (subnet.Match(pnode->addr)) {
+            pnode->fDisconnect = true;
```
I think this belongs in another PR, but happy to do it :-)

Sjors commented:
re-tACK c2e04d3

laanwj commented:
re-utACK 18185b57c32d0a43afeca4c125b9352c692923e9
